File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos13/Alg13.rb
=begin
♦ Algoritmo 13: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Duplamente Ligada/Encadeada: Doubly Linked List (Lista Duplamente Ligada).
=end

class Nodo
  attr_accessor :valor, :anterior, :proximo

  def initialize(valor)
    @valor = valor
    @anterior = nil
    @proximo = nil
  end
end

class ListaDuplamenteLigada
  def initialize
    @primeiro = nil
    @ultimo = nil
    @tamanho = 0
  end

  def add_list(valor)
    novo_nodo = Nodo.new(valor)
    if @primeiro.nil?
      @primeiro = novo_nodo
      @ultimo = novo_nodo
    else
      novo_nodo.anterior = @ultimo
      @ultimo.proximo = novo_nodo
      @ultimo = novo_nodo
    end
    @tamanho += 1
  end

  def remove_list(valor)
    nodo_atual = @primeiro
    while nodo_atual
      if nodo_atual.valor == valor
        if nodo_atual == @primeiro
          @primeiro = nodo_atual.proximo
          @primeiro.anterior = nil if @primeiro
        elsif nodo_atual == @ultimo
          @ultimo = nodo_atual.anterior
          @ultimo.proximo = nil if @ultimo
        else
          anterior = nodo_atual.anterior
          proximo = nodo_atual.proximo
          anterior.proximo = proximo
          proximo.anterior = anterior
        end
        @tamanho -= 1
        return
      end
      nodo_atual = nodo_atual.proximo
    end
  end

  def size
    @tamanho
  end

  def show_list
    nodo_atual = @primeiro
    while nodo_atual
      print "#{nodo_atual.valor} "
      nodo_atual = nodo_atual.proximo
    end
    puts
  end
end

lista = ListaDuplamenteLigada.new
lista.add_list(10)
lista.add_list(20)
lista.add_list(30)

puts "Tamanho da lista: #{lista.size}"
print "Elementos da lista: "
lista.show_list

lista.remove_list(20)

puts "Tamanho da lista após remoção: #{lista.size}"
print "Elementos da lista após remoção: "
lista.show_list

File: Estruturas de Dados e Complexidade de Algoritmos/Algoritmos13/Alg13.cpp
/*
♦ Algoritmo 13: Estrutura de Dados Linear Dinâmica (não Estática) de Lista Duplamente Ligada/Encadeada: Doubly Linked List (Lista Duplamente Ligada).
*/

#include <iostream>

template <typename T>
class Nodo {
public:
    T valor;
    Nodo<T>* anterior;
    Nodo<T>* proximo;

    Nodo(T valor) : valor(valor), anterior(nullptr), proximo(nullptr) {}
};

template <typename T>
class ListaDuplamenteLigada {
private:
    Nodo<T>* primeiro;
    Nodo<T>* ultimo;
    int tamanho;

public:
    ListaDuplamenteLigada() : primeiro(nullptr), ultimo(nullptr), tamanho(0) {}

    void addList(T valor) {
        Nodo<T>* novoNodo = new Nodo<T>(valor);
        if (primeiro == nullptr) {
            primeiro = novoNodo;
            ultimo = novoNodo;
        } else {
            novoNodo->anterior = ultimo;
            ultimo->proximo = novoNodo;
            ultimo = novoNodo;
        }
        tamanho++;
    }

    void removeList(T valor) {
        Nodo<T>* nodoAtual = primeiro;
        while (nodoAtual != nullptr) {
            if (nodoAtual->valor == valor) {
                if (nodoAtual == primeiro) {
                    primeiro = nodoAtual->proximo;
                    if (primeiro != nullptr) {
                        primeiro->anterior = nullptr;
                    }
                } else if (nodoAtual == ultimo) {
                    ultimo = nodoAtual->anterior;
                    if (ultimo != nullptr) {
                        ultimo->proximo = nullptr;
                    }
                } else {
                    Nodo<T>* anterior = nodoAtual->anterior;
                    Nodo<T>* proximo = nodoAtual->proximo;
                    anterior->proximo = proximo;
                    proximo->anterior = anterior;
                }
                tamanho--;
                delete nodoAtual;
                return;
            }
            nodoAtual = nodoAtual->proximo;
        }
    }

    int size() {
        return tamanho;
    }

    void showList() {
        Nodo<T>* nodoAtual = primeiro;
        while (nodoAtual != nullptr) {
            std::cout << nodoAtual->valor << " ";
            nodoAtual = nodoAtual->proximo;
        }
        std::cout << std::endl;
    }

    ~ListaDuplamenteLigada() {
        Nodo<T>* nodoAtual = primeiro;
        while (nodoAtual != nullptr) {
            Nodo<T>* proximo = nodoAtual->proximo;
            delete nodoAtual;
            nodoAtual = proximo;
        }
    }
};

int main() {
    ListaDuplamenteLigada<int> lista;
    lista.addList(10);
    lista.addList(20);
    lista.addList(30);

    std::cout << "Tamanho da lista: " << lista.size() << std::endl;
    std::cout << "Elementos da lista: ";
    lista.showList();

    lista.removeList(20);

    std::cout << "Tamanho da lista após remoção: " << lista.size() << std::endl;
    std::cout << "Elementos da lista após remoção: ";
    lista.showList();

    return 0;
}
